src/map/unit.cpp

```diff
@@ -1404,6 +1407,72 @@ int unit_warp(struct block_list *bl,short m,short x,short y,clr_type type)
 	return 0;
 }
 
+/**
+ * Updates the walkpath of a unit to end after 0.5-1.5 cells moved
+ * Sends required packet for proper display on the client using subcoordinates
+ * @param bl: Object to stop walking
+ */
+void unit_stop_walking_soon(struct block_list& bl)
+{
+	struct unit_data* ud = unit_bl2ud(&bl);
+
+	if (ud == nullptr) return;
+
+	if (ud->walktimer != INVALID_TIMER && ud->walkpath.path_pos + 1 < ud->walkpath.path_len) {
+		const struct TimerData* td = get_timer(ud->walktimer);
+
+		if (td == nullptr) return;
+
+		// Get how much percent we traversed on the timer
+		double cell_percent = 1 - ((double)DIFF_TICK(td->tick, gettick()) / (double)td->data);
+
+		short ox = bl.x, oy = bl.y; // Remember original x and y coordinates
+		short path_remain = 1; // Remaining path to walk
+
+		if (cell_percent > 0 && cell_percent < 1) {
+			// Set subcell coordinates according to timer
+			// This gives a value between 8 and 39
+			ud->sx = (unsigned char)(24.0 + dirx[ud->walkpath.path[ud->walkpath.path_pos]] * 16.0 * cell_percent);
+			ud->sy = (unsigned char)(24.0 + diry[ud->walkpath.path[ud->walkpath.path_pos]] * 16.0 * cell_percent);
+			// 16-31 reflect sub position 0-15 on the current cell
+			// 8-15 reflect sub position 8-15 at -1 main coordinate
+			// 32-39 reflect sub position 0-7 at +1 main coordinate
+			if (ud->sx < 16 || ud->sy < 16 || ud->sx > 31 || ud->sy > 31) {
+				path_remain = 2;
+				if (ud->sx < 16) bl.x--;
+				if (ud->sy < 16) bl.y--;
+				if (ud->sx > 31) bl.x++;
+				if (ud->sy > 31) bl.y++;
+			}
+			ud->sx = ud->sx % 16;
+			ud->sy = ud->sy % 16;
+		}
+		else if (cell_percent >= 1) {
+			// Assume exactly one cell moved
+			bl.x = bl.x + dirx[ud->walkpath.path[ud->walkpath.path_pos]];
+			bl.y = bl.y + diry[ud->walkpath.path[ud->walkpath.path_pos]];
+			path_remain = 2;
+		}
+		// Shorten walkpath
+		if (ud->walkpath.path_pos + path_remain < ud->walkpath.path_len) {
+			ud->walkpath.path_len = ud->walkpath.path_pos + path_remain;
```

Comment:
Here you also need to reduce the walktimer, so that it is triggered earlier.

Reply:
Nothing changes with the timer. The monster just continues moving server-sided.